### PR TITLE
Script to generate windows installer

### PR DESCRIPTION
**Pull request by BenWiederhake**
tl;dr: Windows installer. Some questions at the bottom.

Since this is a major change ("13 changed files with 411 additions and 620 deletions"), and the last iteration discovered lots of things that needed improvement, I think it's a better idea to first ask for more feedback.  Again.

What this does is simple to say: ./mkwindows.sh generates a fresh, redistributable Windows installer from the current workdir.

Potential downsides:

- It's not as well-tested as Eion's hand-tested build.
- The third-party libraries about factor ten bigger, each. Horrible!
- The build takes painstakingly long, around 5 minutes, mostly due to tgl.
- It has to touch tgl's internal build system by hand. But Eion had to do that already each time, so this is still faster and more robust.
- Building is Debian-specific, or at least probably difficult to do on Ubuntu.  It's impossible to do on Ubuntu Trusty, which is Travis' newest image (ugh!).  I set up a virgin Debian VM and built everything from scratch to make sure that my instructions in the README are complete.  It's hopefully more portable than our previous "Eion-only" build.
- Cross-compiling libwebp wasn't easy, and is probably fragile.  Static linking reduces risk of breaking during runtime, and there's a simple switch at the beginning of the script to turn it off.  So if it breaks for a user during build-time, they can circumvent it – for the cost of stickers.
- Cross-compiling sucks, so I didn't get around yet to cross-build libpng yet.  Receving pngs does work though.

Things that I improved since last time:

- MinGW includedir: at least make it a Magic Constant
- Instead of moving around the `auto` files, it should move around `bin/{generate,tl-parser}`
- libwebp version-in-filename
- Fix glibc vs glib
- Document new `#if GLIB_VERSION()` usage

Things I'm still looking at:

- Stickers don't work, so something went wrong with libwebp.  Sigh.
- Why do we need libpng again?
- If it's useful, I should cross-compile it, too.

Half-questions:

- Build both 32-bit and 64-bit?  → No, too much work, don't really need it.
- Move win-specific `make` targets to own file?  → I don't really see the point.  Then I'd need to think about where to draw the line.

### CODE REVIEW

**Comment by EionRobb** [comment]
I quite like it, other than I can't compile on windows any more :) 
But if it gets more regular releases out the door, that's great.

**Comment by BenWiederhake** [comment]
Not more releases per se, but with this I can easily tack the windows installer to a release because I can build it by saying `./mkwindows.sh`.

**Comment by BenWiederhake** [comment]
Here's an example build that should work fine:
[installer.zip](https://github.com/majn/telegram-purple/files/1881951/installer.zip)

**Comment by vn971** [comment]
Considering it gives a "zip" as output, might as well use something like p7zip to produce a self-extracting exe. It works on Windows, too, because 7z self-extraction is basically concatenating a static exe "header" (raw bytes) with the compressed data.

Not sure if you can specify default extraction path with this approach.

**Comment by BenWiederhake** [comment]
@vn971: The installer is an exe.  If you had taken a look at the zip itself (now offline), you would have noticed that it's just the installer, zipped, because GitHub doesn't want executables in the attachement system.

**Comment by BenWiederhake** [comment]
If everything goes right, namely:

- [ ] someone can reproduce the installer, and
- [ ] there's no additional bugs in the resulting libtelegram.dll

then I guess I will merge it in the next few days or weeks.

Here's how you can reproduce it:
- Download: `git clone --recursive -b dev-windows-installer https://github.com/BenWiederhake/telegram-purple.git`
- Install pre-requisites and build, by following https://github.com/BenWiederhake/telegram-purple/tree/dev-windows-installer#building-the-windows-installer

**Comment by BenWiederhake** [comment]
This compression applies to the *installer*, not to the installed files (.dll and .png).  Still, it may save some bandwidth when downloading it.  Some numbers:

The first dev-windows-installer PR: 16,412 KiB  
The current dev-windows-installer PR: 17,016 KiB  
No further changes, just enabling compression: 3,200 KiB

**Comment by BenWiederhake** [comment]
Stripping means a size reduction of the *installed* files (as a side effect, this makes the installer smaller, too).  Only the symbols of dependencies get stripped, as we probably won't ever need those.

Sizes before:  
```
$ du -ch bin/libtelegram.dll contrib/* | sort -h
936K	contrib/libgpg-error-0.dll
1.2M	contrib/libgcc_s_sjlj-1.dll
3.3M	contrib/libwebp-7.dll
4.6M	bin/libtelegram.dll
6.5M	contrib/libgcrypt-20.dll
17M	total
```

Size after:  
```
$ du -ch bin/libtelegram.dll contrib/* | sort -h
100K	contrib/libgcc_s_sjlj-1.dll
108K	contrib/libgpg-error-0.dll
400K	contrib/libwebp-7.dll
664K	contrib/libgcrypt-20.dll
4.6M	bin/libtelegram.dll
5.9M	total
```

Current installer size: 1,692 KiB.  That's much more reasonable.  I think I'll look into #433 next.

**Comment by BenWiederhake** [comment]
Great, someone reproduced the installer on a different machine.  Now I need some Windows users to check whether this installer causes any regressions (= breaking things that used to work).

[Here's the installer built from this branch.  It's zipped because GitHub doesn't like `.exe` files.](https://github.com/majn/telegram-purple/files/1886330/telegram-purple-1.3.1.g6727feef0d.exe.zip)

Please note: Yes, tgp is buggy.  I'm only interested in *new* bugs introduced by this process.  I'm already looking into the other bugs.

**Comment by EionRobb** [comment]
Please don't strip the dll's, it makes looking for crashes via Pidgin.RPT harder :(

**Comment by EionRobb** [comment]
Enabling compression will get the installer flagged for false positives more often.

**Comment by BenWiederhake** [comment]
Then I'll undo the stripping.  But since such a large installer has many other disadvantages, I think we should keep compression.

Nice to see you back!  May I ask why you deleted your Telegram account?  Answering in private or just not answering is okay, too.

**Comment by BenWiederhake** [comment]
Do you have any pointers why stripping gcrypt makes it harder to look for crashes?  I'm not stripping libtelegram.dll, only the other dlls.  Is that still relevant?

**Comment by EionRobb** [comment]
@BenWiederhake haven't deleted my account, just haven't logged into it for a while - too many assert()s! :)

> Do you have any pointers why stripping gcrypt makes it harder to look for crashes?
You're assuming that you're the only plugin that uses gcrypt?  Stripped binaries (even on dependencies) make the stack trace nearly useless on Windows; function calls that come from the stripped binaries aren't informative, or the backtrace will say that the segfault happened in the library, rather than the callback that the library makes.

> But since such a large installer has many other disadvantages, I think we should keep compression.
Sure, just be prepared for the installer to regularly be flagged as a false-positive virus/trojan.  After several years of having that issue with the Skype plugin installer, I gave up and let the downloading client just do gzip content-encoding downloading to save on download size.

**Comment by BenWiederhake** [comment]
Hmm.  I'm afraid this makes sense.  I'll remove the stripping, and document that the compression might cause problems.

**Comment by BenWiederhake** [comment]
One week without complaints, so I'll push it to dev-1.4.0.

[Here's the installer built from this branch.](https://github.com/majn/telegram-purple/files/1912407/telegram-purple-1.3.1.g23f4eaeebc.exe.zip)

Please note:
- It's zipped because GitHub doesn't like .exe files.
- I don't think that the LZMA compression will cause it to be flagged as a virus, at least there's [66 scanners which don't complain about it](https://www.virustotal.com/de/file/21b4511d4476da1d0ea50f268579112ae362374465d5ff90ddb986e5f9f1794c/analysis/1523794364/).
- This is not 1.4.0.  That version will only appear in the future.
- tgp still is buggy.  Sorry.
- This issue and all related ones only close when 1.4.0 gets released.  Again, that's in the future, not the past.

**Comment by himselfv** [comment]
It also asks me to
```
You're missing the file /usr/i686-w64-mingw32/bin/libgcrypt-20.dll
[ ... ]
Also, please file a bug report *EVEN IF THIS SOLVES THE ISSUE*,
as it should have been detected earlier.
```
It had been detected earlier but since under cygwin mingw places all those files for some reason in ``/usr/MINGW_TARGET/sys-root/mingw`` instead of simply in ``/usr/MINGW_TARGET``, I've adjusted the MINGW_BASE path in ``mkwindows.sh``. Everything else until that moment used the adjusted path, but not this.

**Comment by BenWiederhake** [comment]
I cannot find any occurrences of `setopt -e`.  I'm not sure what you mean.  Also the `configure` script is auto-generated, so I hope that a grave mistake like setting `-e` but ignoring exit codes doesn't happen.  Or did your `./configure` return some exit code which causes `./mkwindows.sh` to stop?  This would be intentional, as this means that `./configure` encountered a fatal error.

Regarding your other issue: Oh!  I never considered that someone might want to build it under Windows.  I'm glad to hear that this already mostly works – although the cross-build overhead is probably not needed.  You will need to adapt the paths in the `Makefile`, lines 278, 281, and 284.  I'm not sure yet how to fix this more nicely.  In the meantime, this has to suffice.
PRs welcome! ;)

**Comment by himselfv** [comment]
https://github.com/majn/telegram-purple/blob/dev-1.4.0/mkwindows.sh - right there under copyright. ``mkwindows.sh`` seems to rely on it: if I remove it, ``mkwindows.sh`` continues even in the face of some errors. Meanwhile it calls ``configure``, which does not do ``setopt -e`` itself, but inherits the ``-e`` mode (maybe that shouldn't happen? cygwin thing?). ``configure`` does not expect that mode; some of the commands it calls return non-0 and everything aborts.
Thanks! :)

**Comment by BenWiederhake** [comment]
> but inherits the -e mode

Oh!  What?  That's news to me.  No, that's not intended, and that's the fault.

If I insert "false" as the first command in configure (or `lkjhgliugvfodsz` which hopefully isn't a real program), `./mkwindows` just continues running.  Sure, it complains about `./configure: line 15: kjfvkjvckhgv: command not found`, but then continues.

Can you confirm that this is *not* the behavior you see with Cygwin's shell?

If so, then it sounds like a bug (or at least a quirk) of Cygwin's shell to me.

**Comment by himselfv** [comment]
Yes, that's what it did; apparently, I brought this upon myself by ``export SHELLOPTS`` in ``.bashrc``. Sorry!

**Comment by BenWiederhake** [comment]
Thanks for the feedback! :)

To everyone from the future: If there's any problem with `./mkwindows.sh`, please open a **new ticket**.

**Comment by BenWiederhake** [comment]
Because [things](https://github.com/majn/telegram-purple/issues/478#issuecomment-462051733) happened, here's [another installer built from this branch](https://github.com/majn/telegram-purple/files/2896160/telegram-purple-1.3.1%2Bg5b10181194.exe.zip).

Please note:

- It's zipped because GitHub doesn't like .exe files.
- I don't think that the LZMA compression will cause it to be flagged as a virus.
- This is not 1.4.0. That version will only appear in the future.
- tgp still is buggy. Sorry.
- The related issues only close when 1.4.0 gets released. Again, that's in the future, not the past.

**Comment by BenWiederhake** [comment]
Should be fixed in [v1.4.1](https://github.com/majn/telegram-purple/releases/tag/v1.4.1).

I will remove the above zip'ed installers soon.
